homeworks/03-data-structures/src/merge-sort.ts

```diff
@@ -0,0 +1,35 @@
+type CompareFunction<T> = (a: T, b: T) => number
+
+export const mergeSort = <T>(array: readonly T[], compareFunction: CompareFunction<T>): readonly T[] => {
+    function merge(a: readonly T[], b: readonly T[]): readonly T[] {
+        const results = []
+        const positions = { a: 0, b: 0 }
+
+        // A more elegant recursive solution was overflowing the call stack
```

Comment:
What is the more elegant solution :)?

Reply:
`function merge(a: readonly T[], b: readonly T[]): readonly T[] {
        if (a.length === 0) {
            return b
        } else if (b.length === 0) {
            return a
        } else {
            const [ah, at] = [a[0], a.slice(1)]
            const [bh, bt] = [b[0], b.slice(1)]
            const comparison = compareFunction(ah, bh)
            return comparison < 0 ? [ah, ...merge(at, b)] : [bh, ...merge(a, bt)]
        }
    }`